### PR TITLE
fix bugs in KTO implementation

### PR DESCRIPTION
**Pull request by kawine**
Hi there, KTO author here with a PR. Current KTO implementation puts the mismatched examples (x,y') used to calculate the KL term and the matched examples (x,y) from the data into the same stream. This causes two problems:

- If there are no mismatched examples in a batch, you can NaN for the KL term and a NaN for the loss. This is not fully fixed by increasing batch size (as suggested [here](https://github.com/huggingface/trl/pull/1181))-- you can get a KL estimate that is not NaN but still poor because it is made with so 1 or 2 examples.

- The number of individual examples != the batch size. For example, if you have a batch of 32, depending on how many mismatched examples used to calculate the KL term there are, you can end up with 10 positive/desirables and 2 negative/undesirable losses; 0 positive and 8 negative, etc. This leads to high-variance updates.

(also fixed a minor issue that caused problems with running KTO with accelerate)

### CODE REVIEW

**Comment by kawine** [comment]
cc @kashif @lewtun

**Comment by kashif** [comment]
thanks @kawine having a look now!

**Review by kashif — APPROVED** [review]

File: examples/scripts/kto.py [hunk 1]
```diff
@@ -15,22 +15,41 @@
...
     --max_steps 1000 \
-    --report_to "wandb" \
-    --gradient_checkpointing True  \
-    --output_dir="./test" \
-    --use_peft True \
-    --lora_r 64 \
-    --lora_alpha 16 \
-    --evaluation_strategy "steps" \
-    --logging_first_step True \
+    --learning_rate 1e-3 \
```
Comment (lewtun) [inline examples/scripts/kto.py:23]: For my general understanding, are large learning rates like this the typical values for KTO runs? I noticed in your [config](https://github.com/ContextualAI/HALOs/blob/f9f78263cb7aa70809a87f9f74482dfd09f8c60b/config/config.yaml#L55) the you have 5e-7 which is closer to what I expect.

We aim for these simple examples to "just work", so having a good default here would be great

File: examples/scripts/kto.py [hunk 2]
```diff
@@ -15,22 +15,41 @@
...
+    --bf16 \
+    --logging_first_step \
+    --no_remove_unused_columns
+
+# peft:
+python examples/scripts/kto.py \
+    --model_name_or_path=gpt2 \
+    --per_device_train_batch_size 4 \
+    --max_steps 1000 \
+    --learning_rate 1e-3 \
```
Comment (lewtun) [inline examples/scripts/kto.py:39]: Does KTO with LoRA work better with 10x larger learning rates than full training? If so, I would increase it here

File: trl/trainer/kto_trainer.py [hunk 3]
```diff
@@ -312,25 +312,50 @@ def make_inputs_require_grad(module, input, output):
...
+
+        # merge the datasets
+        train_dataset = concatenate_datasets([train_dataset, train_KL_dataset], axis=1)
+        eval_dataset = concatenate_datasets([eval_dataset, eval_KL_dataset], axis=1)
+
+        desirable = train_dataset.filter(lambda x: x["label"])
+        undesirable = train_dataset.filter(lambda x: not x["label"])
+
+        if (len(desirable) != len(undesirable)):
+            des_weight_lower_bound = (len(undesirable) * self.undesirable_weight / len(desirable)) * 1
```
Comment (lewtun) [inline trl/trainer/kto_trainer.py:336]: Let's add this for clarity:

```suggestion
            # The lower and upper bounds come from Eq. (8) of https://arxiv.org/abs/2402.01306
            des_weight_lower_bound = (len(undesirable) * self.undesirable_weight / len(desirable)) * 1
```

File: trl/trainer/kto_trainer.py [hunk 4]
```diff
@@ -735,46 +810,62 @@ def get_batch_logps(
...
+        KL_logps = self.get_batch_logps(
+            KL_logits,
+            batch["KL_completion_labels"],
+            average_log_prob=False,
+            is_encoder_decoder=self.is_encoder_decoder,
+            label_pad_token_id=self.label_pad_token_id,
+        )
 
-        train_label = [batch["label"][i] for i in target_indicies]
+        assert completion_logps.shape[0] == len(batch['label'])
```
Comment (lewtun) [inline trl/trainer/kto_trainer.py:857]: Let's  raise a `ValueError` here with a nice error message :)

**Review by lewtun — APPROVED** [review]
Thank you very much for all these fixes @kawine - the PR looks great 🔥 !

I've left a few nits, but apart from that LGTM once the style checks pass

File: trl/trainer/kto_trainer.py [hunk 5]
```diff
@@ -27,7 +27,7 @@
 import torch.nn as nn
 import torch.nn.functional as F
 from accelerate.utils import is_deepspeed_available, tqdm
-from datasets import Dataset, interleave_datasets
+from datasets import Dataset, interleave_datasets, concatenate_datasets
```
Comment (kashif) [inline trl/trainer/kto_trainer.py:30]:
```suggestion
from datasets import Dataset, concatenate_datasets, interleave_datasets
```

File: trl/trainer/kto_trainer.py [hunk 6]
```diff
@@ -312,25 +312,50 @@ def make_inputs_require_grad(module, input, output):
...
-            self.shuffle_completion, batched=True, batch_size=128, remove_columns=columns_to_remove
-        )
-        train_dataset = train_dataset.map(self.tokenize_row)
+        # get KL datasets
+        train_KL_dataset = train_dataset.map(self.get_KL_dataset, batched=True, batch_size=1000)
+        if eval_dataset is not None:
+            eval_KL_dataset = eval_dataset.map(self.get_KL_dataset, batched=True, batch_size=1000)
 
+        # tokenize the datasets
+        train_dataset = train_dataset.map(lambda row: self.tokenize_row(row, prefix=""), remove_columns=train_dataset.column_names)
```
Comment (kashif) [inline trl/trainer/kto_trainer.py:321]:
```suggestion
        train_dataset = train_dataset.map(
            lambda row: self.tokenize_row(row, prefix=""), remove_columns=train_dataset.column_names
        )
```

File: trl/trainer/kto_trainer.py [hunk 7]
```diff
@@ -312,25 +312,50 @@ def make_inputs_require_grad(module, input, output):
...
+        train_KL_dataset = train_KL_dataset.map(lambda row: self.tokenize_row(row, prefix="KL_"), remove_columns=train_KL_dataset.column_names)
```
Comment (kashif) [inline trl/trainer/kto_trainer.py:322]:
```suggestion
        train_KL_dataset = train_KL_dataset.map(
            lambda row: self.tokenize_row(row, prefix="KL_"), remove_columns=train_KL_dataset.column_names
        )
```

File: trl/trainer/kto_trainer.py [hunk 8]
```diff
@@ -312,25 +312,50 @@ def make_inputs_require_grad(module, input, output):
...
         if eval_dataset is not None:
-            columns_to_remove = [
-                col for col in eval_dataset.column_names if col not in ["prompt", "completion", "label"]
-            ]
-            eval_dataset = eval_dataset.map(
-                self.shuffle_completion, batched=True, batch_size=128, remove_columns=columns_to_remove
-            )
-            eval_dataset = eval_dataset.map(self.tokenize_row)
+            eval_dataset = eval_dataset.map(lambda row: self.tokenize_row(row, prefix=""), remove_columns=eval_dataset.column_names)
+            eval_KL_dataset = eval_KL_dataset.map(lambda row: self.tokenize_row(row, prefix="KL_"), remove_columns=eval_KL_dataset.column_names)
```
Comment (kashif) [inline trl/trainer/kto_trainer.py:326]:
```suggestion
            eval_dataset = eval_dataset.map(
                lambda row: self.tokenize_row(row, prefix=""), remove_columns=eval_dataset.column_names
            )
            eval_KL_dataset = eval_KL_dataset.map(
                lambda row: self.tokenize_row(row, prefix="KL_"), remove_columns=eval_KL_dataset.column_names
            )
```

File: trl/trainer/kto_trainer.py [hunk 9]
```diff
@@ -312,25 +312,50 @@ def make_inputs_require_grad(module, input, output):
...
+
+        # merge the datasets
+        train_dataset = concatenate_datasets([train_dataset, train_KL_dataset], axis=1)
+        eval_dataset = concatenate_datasets([eval_dataset, eval_KL_dataset], axis=1)
+
+        desirable = train_dataset.filter(lambda x: x["label"])
+        undesirable = train_dataset.filter(lambda x: not x["label"])
+
+        if (len(desirable) != len(undesirable)):
```
Comment (kashif) [inline trl/trainer/kto_trainer.py:335]:
```suggestion
        if len(desirable) != len(undesirable):
```

File: trl/trainer/kto_trainer.py [hunk 10]
```diff
@@ -312,25 +312,50 @@ def make_inputs_require_grad(module, input, output):
...
+            des_weight_lower_bound = (len(undesirable) * self.undesirable_weight / len(desirable)) * 1
+            des_weight_upper_bound = (len(undesirable) * self.undesirable_weight / len(desirable)) * 1.33
+            und_weight_lower_bound = (len(desirable) * self.desirable_weight / len(undesirable)) / 1.33
+            und_weight_upper_bound = (len(desirable) * self.desirable_weight / len(undesirable)) / 1
+
+            des_weight_in_range = (des_weight_lower_bound <= self.desirable_weight <= des_weight_upper_bound)
+            und_weight_in_range = (und_weight_lower_bound <= self.undesirable_weight <= und_weight_upper_bound)
```
Comment (kashif) [inline trl/trainer/kto_trainer.py:342]:
```suggestion
            des_weight_in_range = des_weight_lower_bound <= self.desirable_weight <= des_weight_upper_bound
            und_weight_in_range = und_weight_lower_bound <= self.undesirable_weight <= und_weight_upper_bound

```

File: trl/trainer/kto_trainer.py [hunk 11]
```diff
@@ -437,14 +458,24 @@ def get_train_dataloader(self) -> DataLoader:
...
+                reference_completion_logp = self.accelerator.gather_for_metrics(reference_completion_logp)
+                reference_completion_logps.append(reference_completion_logp.cpu())
+
+                reference_KL_logp = self.accelerator.gather_for_metrics(reference_KL_logp)
+                reference_KL_logps.append(reference_KL_logp.cpu())
+
+            self.train_dataset = self.train_dataset.add_column(name="reference_logps", 
+                                                   column=torch.cat(reference_completion_logps).float().numpy())
+            self.train_dataset = self.train_dataset.add_column(name="reference_KL_logps", 
+                                                   column=torch.cat(reference_KL_logps).float().numpy())
```
Comment (kashif) [inline trl/trainer/kto_trainer.py:477]:
```suggestion
            self.train_dataset = self.train_dataset.add_column(
                name="reference_logps", column=torch.cat(reference_completion_logps).float().numpy()
            )
            self.train_dataset = self.train_dataset.add_column(
                name="reference_KL_logps", column=torch.cat(reference_KL_logps).float().numpy()
            )
```

File: trl/trainer/kto_trainer.py [hunk 12]
```diff
@@ -476,14 +507,22 @@ def get_eval_dataloader(self, eval_dataset: Optional[Dataset] = None) -> DataLoa
...
 
-            all_reference_logps = torch.cat(reference_logps).float().numpy()
-            eval_dataset = eval_dataset.add_column(name="reference_logps", column=all_reference_logps)
+                reference_KL_logp = self.accelerator.gather_for_metrics(reference_KL_logp)
+                reference_KL_logps.append(reference_KL_logp.cpu())
+
+            eval_dataset = eval_dataset.add_column(name="reference_logps", 
+                                                   column=torch.cat(reference_completion_logps).float().numpy())
+            eval_dataset = eval_dataset.add_column(name="reference_KL_logps", 
+                                                   column=torch.cat(reference_KL_logps).float().numpy())
```
Comment (kashif) [inline trl/trainer/kto_trainer.py:525]:
```suggestion
            eval_dataset = eval_dataset.add_column(
                name="reference_logps", column=torch.cat(reference_completion_logps).float().numpy()
            )
            eval_dataset = eval_dataset.add_column(
                name="reference_KL_logps", column=torch.cat(reference_KL_logps).float().numpy()
            )
```

File: trl/trainer/kto_trainer.py [hunk 13]
```diff
@@ -500,38 +539,71 @@ def compute_reference_log_probs(self, padded_batch: Dict) -> Dict:
...
                 else:
-                    all_logits = self.ref_model(
+                    completion_logits = self.ref_model(
                         padded_batch["completion_input_ids"], attention_mask=padded_batch["completion_attention_mask"]
                     ).logits
 
-        return self.get_batch_logps(
-            all_logits,
+                    KL_logits = self.ref_model(
+                        padded_batch["KL_completion_input_ids"], attention_mask=padded_batch["KL_completion_attention_mask"]
```
Comment (kashif) [inline trl/trainer/kto_trainer.py:586]:
```suggestion
                        padded_batch["KL_completion_input_ids"],
                        attention_mask=padded_batch["KL_completion_attention_mask"],
```

File: trl/trainer/kto_trainer.py [hunk 14]
```diff
@@ -735,46 +810,62 @@ def get_batch_logps(
...
         if self.is_encoder_decoder:
-            all_logits = model(
+            with torch.no_grad():
+                KL_logits = model(
+                    batch["KL_prompt_input_ids"],
+                    attention_mask=batch["KL_prompt_attention_mask"],
+                    decoder_input_ids=batch.get("KL_completion_decoder_input_ids"),
+                    labels=batch["KL_completion_labels"],
+                ).logits
+
```
Comment (kashif) [inline trl/trainer/kto_trainer.py:822]:
```suggestion

```

File: trl/trainer/kto_trainer.py [hunk 15]
```diff
@@ -886,18 +972,15 @@ def get_batch_loss_metrics(
...
         )
 
-        reward_accuracies = (chosen_rewards.mean() > rejected_rewards.mean()).float()
-
         prefix = "eval_" if train_eval == "eval" else ""
         metrics[f"{prefix}rewards/chosen"] = chosen_rewards.mean().cpu()
         metrics[f"{prefix}rewards/rejected"] = rejected_rewards.mean().cpu()
-        metrics[f"{prefix}rewards/accuracies"] = reward_accuracies.cpu()
-        metrics[f"{prefix}rewards/margins"] = (chosen_rewards.mean() - rejected_rewards.mean()).cpu()
+        metrics[f"{prefix}rewards/margins"] = (chosen_rewards.mean().nan_to_num(0) - rejected_rewards.mean().nan_to_num(0)).cpu()
```
Comment (kashif) [inline trl/trainer/kto_trainer.py:978]:
```suggestion
        metrics[f"{prefix}rewards/margins"] = (
            chosen_rewards.mean().nan_to_num(0) - rejected_rewards.mean().nan_to_num(0)
        ).cpu()
```

File: trl/trainer/kto_trainer.py [hunk 16]
```diff
@@ -886,18 +972,15 @@ def get_batch_loss_metrics(
...
         metrics[f"{prefix}logps/rejected"] = policy_rejected_logps.detach().mean().cpu()
         metrics[f"{prefix}logps/chosen"] = policy_chosen_logps.detach().mean().cpu()
         metrics[f"{prefix}logits/rejected"] = policy_rejected_logits.detach().mean().cpu()
         metrics[f"{prefix}logits/chosen"] = policy_chosen_logits.detach().mean().cpu()
-        metrics[f"{prefix}kl"] = kl.item()
+        metrics[f"{prefix}kl"] = kl.item() # has already been gathered in kto_loss
```
Comment (kashif) [inline trl/trainer/kto_trainer.py:983]:
```suggestion
        metrics[f"{prefix}kl"] = kl.item()  # has already been gathered in kto_loss
```

**Comment by kashif** [comment]
@kawine  feel free to just apply the suggestions which are stylistic fixes to get the CI to pass (apart from the `assert` issue flagged above)

File: examples/scripts/kto.py (hunk 1 above)
Comment (kawine) [reply examples/scripts/kto.py:23]: I just copied over the suggested commands in scripts/dpo.py and changed dpo -> kto. The command for DPO also seems to use a different learning rate than the 5e-7 that the original DPO uses?

All our experiments in the paper used the same training configs for DPO and KTO, so I'd recommend they be the same here as well (what exactly those configs are gpt2, I don't have a strong opinion on).

File: examples/scripts/kto.py (hunk 2 above)
Comment (kawine) [reply examples/scripts/kto.py:39]: As mentioned above, this was just copy-pasted from scripts/dpo.py. In all our experiments so far, we've used identical training configs for both, usually what the DPO repo recommended.

File: trl/trainer/kto_trainer.py (hunk 4 above)
Comment (kawine) [reply trl/trainer/kto_trainer.py]: done
